hazelcast/sql.py

```diff
@@ -0,0 +1,1593 @@
+            row_page (_SqlPage): The first page of the result. Might be
+                ``None`` if the response only contains the update count.
+            update_count (int): The update count.
+        """
+        with self._lock:
+            if self._closed:
+                # User might be already cancelled it.
+                return
+
+            if row_metadata:
+                # Result contains the row set for the query.
+                # Set the update count to -1.
+                response = _ExecuteResponse(row_metadata, row_page, -1)
+
+                if row_page.is_last:
+                    # This is the last page, close the result.
+                    self._closed = True
+
+                self._execute_response.set_result(response)
+            else:
+                # Result only contains the update count.
+                response = _ExecuteResponse(None, None, update_count)
+                self._execute_response.set_result(response)
+
+                # There is nothing more we can get from the server.
+                self._closed = True
+
+    def __enter__(self):
+        # The execute request is already sent.
+        # There is nothing more to do.
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        # Ignoring the possible exception details
+        # since we close the query regardless of that.
+        self.close().result()
+
+
+class _InternalSqlService(object):
+    """Internal SQL service that offers more public API
+    than the one exposed to the user.
+    """
+
+    def __init__(self, connection_manager, serialization_service, invocation_service):
+        self._connection_manager = connection_manager
+        self._serialization_service = serialization_service
+        self._invocation_service = invocation_service
+
+    def execute(self, sql, *params):
+        """Constructs a statement and executes it.
+
+        Args:
+            sql (str): SQL string.
+            *params: Query parameters.
+
+        Returns:
+            SqlResult: The execution result.
+        """
+        statement = SqlStatement(sql)
+
+        for param in params:
+            statement.add_parameter(param)
+
+        return self.execute_statement(statement)
+
+    def execute_statement(self, statement):
+        """Executes the given statement.
+
+        Args:
+            statement (SqlStatement): The statement to execute.
+
+        Returns:
+            SqlResult: The execution result.
+        """
+
+        # Get a random Data member (non-lite member)
+        connection = self._connection_manager.get_random_connection(True)
+        if not connection:
+            # Either the client is not connected to the cluster, or
+            # there are no data members in the cluster.
+            raise HazelcastSqlError(
+                self.get_client_id(),
+                _SqlErrorCode.CONNECTION_PROBLEM,
+                "Client is not currently connected to the cluster.",
```

Comment:
I still believe it may be misleading, i.e when there is no connection problem -client is connected to a cluster with lite members-; it will say client is not connected to the cluster